From bdcd5d6a0c9ed0b21de33d7bfe34c0f43ced8edd Mon Sep 17 00:00:00 2001
From: Adrien Schildknecht <adrs@fb.com>
Date: Wed, 20 Jan 2021 13:44:15 +0100
Subject: [PATCH] Properly parse IPv6 services in the cli

Summary:
This commit fixes #31.

Eventhough ncrack supports IPv6, the cli interface does not accept IPv6
services because of a parsing issue.
The `-6` flag works fine if a hostname is provided:
  $ ncrack -v -6 ftp://facebook.com
  ftp://2a03:2880:f106:83:face:b00c:0:25de:21 finished.
But the cli rejects the input if we use an IPv6 directly:
  $ ncrack -v -6 [2a03:2880:f106:83:face:b00c:0:25de]:21
  Invalid port number: 2880:f106:83:face:b00c:0:25de]:21
  $ ncrack -v -6 ftp://2a03:2880:f106:83:face:b00c:0:25de
  Invalid port number: 2880:f106:83:face:b00c:0:25de

This diff refactors and changes the way we are parsing services from
the cli to properly handle IPv6 addresses.

Test Plan:
The 2 examples aboves that weren't working:
  $ ./ncrack -v -6 [2a03:2880:f106:83:face:b00c:0:25de]:21
  ftp://2a03:2880:f106:83:face:b00c:0:25de:21 finished.
  $ ./ncrack -v -6 ftp://2a03:2880:f106:83:face:b00c:0:25de
  ftp://2a03:2880:f106:83:face:b00c:0:25de:21 finished.

A small shell script to compare the behavior of the old and new code:
  $ cat test.sh
  cases=(
      '127.0.0.1'
      '127.0.0.1,'
      '127.0.0.1 -p ftp'
      '127.0.0.1,path=/ -p ftp'
      '127.0.0.1:21'
      'ftp://127.0.0.1'
      '203.0.113.0/24'
      '203.0.113.10-20'
      'localhost'
      'ftp://localhost:21'
      '-6 ::1'
      '-6 2001:db8::1'
      '-6 ftp://2001:db8::1'
      '-6 ftp://[2001:db8::1]'
      '-6 [2001:db8::1]:21'
      '-6 ftp://[2001:db8::1]:21'
      '-6 ftp://2001:db8::192.168.0.1'
      '-6 2001:db8::1,path=/ -p ftp'
      '-6 ftp://[2001:db8::1]:21,path=/ -p ftp'
  )
  for args in "${cases[@]}"; do
      printf '%-50s' "$args"
      ncrack $args >/dev/null 2>/dev/null
      echo -en "\t$?"
      ./ncrack $args >/dev/null 2>/dev/null
      echo -e "\t$?"
  done

  $ make && bash test.sh
  ncrack 127.0.0.1                                     	1	1
  ncrack 127.0.0.1,                                    	1	1
  ncrack 127.0.0.1 -p ftp                              	0	0
  ncrack 127.0.0.1,path=/ -p ftp                       	1	1
  ncrack 127.0.0.1:21                                  	0	0
  ncrack ftp://127.0.0.1                               	0	0
  ncrack 203.0.113.0/24                                	1	1
  ncrack 203.0.113.10-20                               	1	1
  ncrack localhost                                     	1	1
  ncrack ftp://localhost:21                            	0	0
  ncrack -6 ::1                                        	1	1
  ncrack -6 2001:db8::1                                	1	1
  ncrack -6 ftp://2001:db8::1                          	1	0
  ncrack -6 ftp://[2001:db8::1]                        	1	0
  ncrack -6 [2001:db8::1]:21                           	1	0
  ncrack -6 ftp://[2001:db8::1]:21                     	1	0
  ncrack -6 ftp://2001:db8::192.168.0.1                	1	0
  ncrack -6 2001:db8::1,path=/ -p ftp                  	1	1
  ncrack -6 ftp://[2001:db8::1]:21,path=/ -p ftp       	1	0

As we can see the new code behave like the old one for IPv4/hostname,
but is also working with valid IPv6 services.
---
 services.cc | 109 +++++++++++++++++++++++++---------------------------
 1 file changed, 52 insertions(+), 57 deletions(-)

diff --git a/services.cc b/services.cc
index 7c2fbde9..39ccf070 100644
--- a/services.cc
+++ b/services.cc
@@ -161,69 +161,64 @@ ts_spec
 parse_services_target(char *const exp) 
 {
   ts_spec temp;
-  size_t name_len, host_len, port_len, options_len, tot_len;
-  char *s, *h, *p;
+  char *s, *h, *p, *o;
+  size_t host_len;
+  int num_colons = 0, lbrackets = 0, rbrackets = 0;
 
-  tot_len = strlen(exp);
   memset(&temp, 0, sizeof(temp));
-  name_len = port_len = 0;
-  p = NULL;
 
+  /* Extract the service name */
   if ((s = strstr(exp, "://"))) {
-    name_len = s - exp;
-    temp.service_name = Strndup(exp, name_len);
-    s += sizeof("://") - 1;
-  } else
-    s = exp;  /* no service name */
-
-  /* case when we have arguments */
-  if ((h = strchr(s, ','))) {
-
-    if ((p = strchr(s, ':')) && (p < h)) {
-      port_len = h - p - 1;
-      temp.portno = Strndup(++p, port_len);
-    } else if (s == exp)  /* neither port nor service name! */
-        fatal("You must specify either a service name or a port (or both): %s",
-            exp);
-
-    if (p) { /* port has been specified */
-      host_len = h - s - port_len - 1;
-      if (s == exp) {
-      /* No service name was provided so we find the default one based on the
-       * port by looking at the supported services table (ServicesTable). */
-        if (!(temp.service_name = port2name(temp.portno)))
-          temp.error = true;
-      }
-    } else {
-      /* port not specified, but to reach here means that sevice-name had
-       * been specified */
-      host_len = h - s;
-    }
-
-    options_len = exp + tot_len - h;
-    temp.service_options = Strndup(++h, options_len);
-
-  } else {  /* case of no arguments */
-    if ((p = strchr(s, ':'))) {
-      port_len = exp + tot_len - p;
-      temp.portno = Strndup(++p, port_len);
-    }
-
-    if (p) {
-      host_len = --p - s;
-      if (s == exp) {
-      /* No service name was provided so we find the default one based on the
-       * port by looking at the supported services table (ServicesTable). */
-         if (!(temp.service_name = port2name(temp.portno)))
-           temp.error = true;
-      }
-    } else {
-      /* only hostname specified (-p option will determine services) */
-      host_len = exp + tot_len - s;
-    }
+      temp.service_name = Strndup(exp, s-exp);
+      s += sizeof("://") - 1;
+  } else {
+      s = exp; /* no service name */
   }
 
-  temp.host_expr = Strndup(s, host_len);
+  /* Determine if it's an IPv6 or not */
+  for (h = s; *s && *s != ','; ++s) {
+    if (*s == ':')
+      ++num_colons;
+    else if (*s == '[')
+      ++lbrackets;
+    else if (*s == ']')
+      ++rbrackets;
+  }
+  if (!(lbrackets == rbrackets && lbrackets <= 1))
+    fatal("Invalid number of brackets:%s", exp);
+
+  /* Extract the host */
+  if (lbrackets) { /* IPv6, potentially with a port */
+    p = strchr(h, ']') + 1;
+    h++;
+    host_len = p-h-1;
+  } else if (num_colons >= 2) { /* IPv6, no port */
+    p = strchrnul(h, ',');
+    host_len = p-h;
+  } else { /* IPv4 or hostname, potentially with a port */
+    p = strchr(h, ':');
+    if (!p)
+      p = strchrnul(h, ',');
+    host_len = p-h;
+  }
+  temp.host_expr = Strndup(h, host_len);
+
+  /* Extract the port */
+  o = strchrnul(s, ',');
+  if (*p == ':')
+    temp.portno = Strndup(p+1, o-p-1);
+
+  /* Extract the options */
+  if (*o == ',')
+    temp.service_options = strdup(o+1);
+
+  if (!temp.service_name && !temp.portno && temp.service_options)
+    fatal("You must specify either a service name or a port (or both): %s",
+        exp);
+  if (!temp.service_name && temp.portno) {
+    if (!(temp.service_name = port2name(temp.portno)))
+      temp.error = true;
+  }
 
   return temp;
 }
